test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -240,49 +244,69 @@ class ApiUsers::PermissionsControllerTest < ActionController::TestCase
 
       assert_not_authorised
     end
 
     should "push permission changes out to apps" do
-      sign_in create(:superadmin_user)
+      application = create(:application)
+      permission = create(:supported_permission, application:)
 
-      application = create(:application, with_non_delegatable_supported_permissions: %w[permission])
-      api_user = create(:api_user)
-      create(:access_token, resource_owner_id: api_user.id, application:)
+      api_user = create(:api_user, with_signin_permissions_for: [application])
+      create(:access_token, application:, resource_owner_id: api_user.id)
 
-      permission = application.supported_permissions.find_by(name: "permission")
+      sign_in create(:superadmin_user)
 
       PermissionUpdater.expects(:perform_on).with(api_user)
 
-      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [permission.id] } }
+      patch :update, params: {
+        api_user_id: api_user,
+        application_id: application,
+        application: { supported_permission_ids: [permission.id] },
+      }
     end
 
-    should "exclude applications with revoked access tokens" do
-      sign_in create(:superadmin_user)
-
+    should "prevent updating permissions if the user has a revoked access token" do
       application = create(:application)
-      api_user = create(:api_user)
+      permission = create(:supported_permission, application:)
+
+      api_user = create(:api_user, with_signin_permissions_for: [application])
+
       create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
 
+      sign_in create(:superadmin_user)
+
       assert_raises(ActiveRecord::RecordNotFound) do
-        patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [] } }
+        patch :update, params: {
+          api_user_id: api_user,
+          application_id: application,
+          application: { supported_permission_ids: [permission] },
+        }
       end
     end
 
-    should "include applications with revoked access tokens when there is at least one non-revoked access token" do
-      sign_in create(:superadmin_user)
+    should "allow updating when the user has revoked access tokens when there is at least one non-revoked access token" do
+      application = create(:application)
+      permission = create(:supported_permission, application:)
 
-      application = create(:application, with_non_delegatable_supported_permissions: %w[permission])
-      api_user = create(:api_user)
+      api_user = create(:api_user, with_signin_permissions_for: [application])
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      sign_in create(:superadmin_user)
       create(:access_token, resource_owner_id: api_user.id, application:, revoked_at: Time.current)
       create(:access_token, resource_owner_id: api_user.id, application:)
```

Comment:
Looks like you're creating the access tokens in two places. I think we just want the second and third ones, but in the position of the first one?

Reply:
Oops, good spot. Have fixed.